src/bt.rs

```diff
@@ -0,0 +1,435 @@
+                    return Err(BinaryTreeError::UnreachableNode(value));
+                }
+            };
+        }
+
+        // Checks whether the node already has a value,
+        // If so, returns an error wrapping the value that could not be inserted.
+        // otherwise, node is available and the new value is inserted.
+        match node {
+            IterNode::Root(_) | IterNode::Inner(Some(_)) => {
+                return Err(BinaryTreeError::InsertNonEmptyNode(value))
+            }
+            IterNode::Inner(empty_node) => {
+                *empty_node = Some(Box::new(Node::new(value)));
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Gets a reference to the value located at the end of the path.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get(&self, path: &Path) -> Option<&V> {
+        let mut node = self;
+        for bit in path {
+            match if !bit { &node.left } else { &node.right } {
+                None => return None,
+                Some(next_node) => node = next_node,
+            };
+        }
+        Some(&node.value)
+    }
+
+    /// Gets a mutable reference to the node located at the end of the path.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get_node(&mut self, path: &Path) -> Option<&mut Node<V>> {
+        let mut node = self;
+        for bit in path {
+            match if !bit {
+                &mut node.left
+            } else {
+                &mut node.right
+            } {
+                None => return None,
+                Some(next_node) => node = next_node,
+            };
+        }
+        Some(node)
+    }
+}
+
+/// Represents an append-only binary tree.
+pub struct BinaryTree<V> {
+    root: SubTree<V>,
+}
+
+impl<V> BinaryTree<V> {
+    /// Inserts the value at the end of the path.
+    ///
+    /// # Returns
+    /// - `Ok(())`, the node was inserted at the end of the path.
+    /// - `Err(InsertNonEmptyNode(value))`, the node already contains a value.
+    /// - `Err(UnreachablePath(value))`, the end of the path is unreachable.
+    ///
+    /// In the error cases, no insertion occurs and the value is returned to its owner.
+    pub fn insert(&mut self, path: &Path, value: V) -> Result<(), BinaryTreeError<V>> {
+        if let Some(node) = &mut self.root {
+            node.insert(path, value)
+        } else if path.is_empty() {
+            self.root = Some(Box::new(Node::new(value)));
+            Ok(())
+        } else {
+            Err(BinaryTreeError::UnreachableNode(value))
+        }
+    }
+
+    /// Gets a reference to the value located at the end of the path.
+    /// [`None`] is returned if no values are found during the traversal of the tree
+    /// along the path.
+    pub fn get(&self, path: &Path) -> Option<&V> {
+        self.root.as_deref().map(|node| node.get(path)).unwrap()
```

Comment:
I think this should be an and_then rather than a map and unwrap. Particularly, `BinaryTree::default().get(&bits!())` should return None rather than panic.

Reply:
good catch